runtime/ftplugin/javascript.vim

```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
 
 setlocal commentstring=//%s
 
-" Change the :browse e filter to primarily show Java-related files.
+" Change the :browse e filter to primarily show JavaScript-related files.
 if has("gui_win32")
-    let  b:browsefilter="Javascript Files (*.js)\t*.js\n" .
-		\	"All Files (*.*)\t*.*\n"
+    let  b:browsefilter="Javascript Files (*.js, *.jsx, *.es, *.es6, *.cjs, *.mjs, *.jsm, *.vue, *.json)\t*.js;*.jsx;*.es;*.es6;*.cjs;*.mjs;*.jsm;*.vue;*.json\n" .
+                \ "All Files (*.*)\t*.*\n"
 endif
-       
-let b:undo_ftplugin = "setl fo< ofu< com< cms<" 
+
+" The following suffixes should be implied when resolving filenames
+setlocal suffixesadd+=.js,.jsx,.es,.es6,.cjs,.mjs,.jsm,.vue,.json
+
+" The following suffixes should have low priority
+"   .snap    jest snapshot
+setlocal suffixes+=.snap
+
+" Prepend node_modules/.bin to $PATH if applicable
+" Allows calling executables installed locally via npm/yarn
+let s:bin_dir = finddir('node_modules/.bin', '.;')->fnamemodify(':p')
+if len(s:bin_dir) && $PATH !~ s:bin_dir
+    let $PATH = s:bin_dir .. ':' .. $PATH
+endif
+unlet s:bin_dir
+
+" Set 'path' to a minimal, non-greedy, value
+" User is expected to augment it with contextually-relevant paths
+setlocal path-=node_modules
+setlocal path-=/usr/include
+setlocal path-=**
+
+" Matchit configuration
+let b:match_words = '\<function\>:\<return\>,'
+            \ .. '\<do\>:\<while\>,'
+            \ .. '\<switch\>:\<case\>:\<default\>,'
+            \ .. '\<if\>:\<else\>,'
+            \ .. '\<try\>:\<catch\>:\<finally\>,'
+            \ .. '<\(\w\+\):</\1>'
+
+" Use eslint for :make if applicable
+if executable('eslint')
+    compiler eslint
+endif
+
+" Use standard for :make if applicable
+if executable('standard')
+    compiler standard
+endif
+
+" Set 'define' to a comprehensive value
+let &define = '\('
+            \ .. '\(^\s*(*async\s\{-}function\|(*function\)'
+            \ .. '\|^\s*\(\*\|static\|async\|get\|set\|\i\{-}\.\)'
+            \ .. '\|^\s*\(\ze\i\{-}\)\(([^)]*).*{$\|\s*[:=,]\)'
+            \ .. '\|^\s*\(export\s\{-}\|export\s\{-}default\s\{-}\)*\(var\|let\|const\|function\)'
+            \ .. '\|\(\<as\>\)'
+            \ .. '\)'
```

Comment:
I'll test this when I get a bit more time but just looking at it for now:

-  the entire pattern doesn't need to be grouped with parens nor does `\<as\>`
- whenever you've used `\{-}` you meant `\+`
- it's not matching class definitions or generator functions

Reply:
- I will look into the parentheses issues.
- No, I meant `\{-1,}`. Will fix this.
- Generator functions are matched but classes are not, I'll make sure it does.

Reply:
>     * I will look into the parentheses issues.
> 
>     * No, I meant `\{-1,}`. Will fix this.
That always matches the same thing as `\+` in the contexts it's currently used and reads a bit like a regex double negative to me.
>     * Generator functions are matched but classes are not, I'll make sure it does.
Sorry I forgot it skips non-identifier chars rather than just whitespace.

If you're working on this now I'll wait for the next update to test it.

Reply:
@dkearns, commit 8f245c29a:

- adds class definitions,
- replaces all `\{-}` with `\+`
- removes some unnecessary parentheses.

FWIW, the value supports the following in its current state:

    function foo
    (function foo
    async function foo
    (async function foo
    function * foo
    var foo
    let foo
    const foo
    static foo
    async foo
    get foo
    set foo
    this.foo
    identifier.foo
    foo()
    foo =
    foo,
    export let foo
    export var foo
    export const foo
    export function foo
    export class foo
    export default let foo
    export default const foo
    export default var foo
    export default function foo
    export default class foo
    import ... as foo
    export ... as foo
    * foo

NOTE: Vim is incapable of finding multiple definitions on the same line.